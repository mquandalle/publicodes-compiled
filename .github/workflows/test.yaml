name: test
on: push
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: oven-sh/setup-bun@v1
      - run: bun install
      - run: bun test
