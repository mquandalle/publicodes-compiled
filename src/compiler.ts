import { walk } from "zimmerframe";
import { builders as b } from "estree-toolkit";
import { generate } from "astring";
import { type ASTPublicodesNode } from "./parser";

export function compile(parsedPublicodes: ASTPublicodesNode) {
  const state = { injectedRuntime: new Set() };
  const esTree = walk(parsedPublicodes, state, {
    publicodes(node, { visit }) {
      return b.objectExpression(
        node.rules.map((ruleNode) =>
          b.property("init", b.literal(ruleNode.name), visit(ruleNode))
        )
      );
    },

    rule(node, { visit }) {
      return b.arrowFunctionExpression([b.identifier("r")], visit(node.value));
    },

    constant(node) {
      return b.literal(node.value);
    },

    reference(node) {
      return b.callExpression(b.identifier("r"), [b.literal(node.name)]);
    },

    operation(node, { visit }) {
      return b.binaryExpression(
        node.operator,
        visit(node.left),
        visit(node.right)
      );
    },

    undefined() {
      return b.identifier("undefined");
    },

    produit(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.assiette,
        right: node.taux,
      });
    },

    unitConversion(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.factor,
        right: node.value,
      });
    },

    barÃ¨me(node, { state, visit }) {
      state.injectedRuntime.add(
        "const evalBareme = (a, t) => t.map((u, i) => ((Math.max(Math.min(a, u.plafond ?? Infinity) - (t[i - 1]?.plafond ?? 0), 0)) * u.taux)).reduce((a, b) => a + b, 0);"
      );
      return b.callExpression(b.identifier("evalBareme"), [
        visit(node.assiette),
        b.arrayExpression(
          node.tranches.map((t) =>
            b.objectExpression(
              ["taux", t.plafond && "plafond"]
                .filter(Boolean)
                .map((attr) =>
                  b.property("init", b.identifier(attr), visit(t[attr]))
                )
            )
          )
        ),
      ]);
    },

    ["une de ces conditions"](node, { visit }) {
      let res = visit(node.value[0]);
      for (let i = 1; i < node.value.length; i++) {
        res = b.logicalExpression("||", res, visit(node.value[i]));
      }
      return res;
    },

    ["toutes ces conditions"](node, { visit }) {
      let res = visit(node.value[0]);
      for (let i = 1; i < node.value.length; i++) {
        res = b.logicalExpression("&&", res, visit(node.value[i]));
      }
      return res;
    },

    variations(node, { visit }) {
      return b.conditionalExpression(
        visit(node.value[0].si),
        visit(node.value[0].alors),
        visit(node.value[1].sinon)
      );
    },

    ["applicable si"](node, { visit }) {
      return b.conditionalExpression(
        visit(node["applicable si"]),
        visit(node["value"]),
        b.identifier("null")
      );
    },
  });

  const jsCodeRules = generate(esTree);
  const injectedRuntime = Array.from(state.injectedRuntime).join(";\n");
  return `(() => {${injectedRuntime}\nreturn ${jsCodeRules}})`;
}
