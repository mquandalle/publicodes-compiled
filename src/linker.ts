import type { ASTNode, ASTPublicodesNode, ASTRuleNode } from "./parser";
import { walk } from "zimmerframe";
import { conversionFactor, inferUnit } from "./units";

type InferedType =
  | { type: "string" | "boolean" }
  | { type: "number"; unit: string };

export function link(parsedRules: ASTPublicodesNode) {
  const inferedUnits = new WeakMap<ASTNode, InferedType>();
  const rewrittenRules = new WeakMap<ASTNode, ASTNode>();
  const availableNames = parsedRules.rules.map((rule) => rule.name);

  function inferRuleUnit(ruleNode: ASTRuleNode) {
    if (rewrittenRules.has(ruleNode)) {
      return rewrittenRules.get(ruleNode);
    }
    return walk(
      ruleNode,
      { enforceUnitStack: [] },
      {
        rule(node, { visit, state }) {
          const allParents = (name) =>
            name
              .split(".")
              .reduce((acc, name) => {
                const last = acc[acc.length - 1];
                return [
                  ...acc,
                  last ? `${last} . ${name.trim()}` : name.trim(),
                ];
              }, [])
              .reverse();

          const rewrittenNode = visit(node.value, {
            ...state,
            parentsNames: allParents(node.name),
          });
          const rewrittenNodeUnit = inferedUnits.get(rewrittenNode);
          inferedUnits.set(node, inferedUnits.get(rewrittenNode));
          const rewrittenRule = {
            ...node,
            value: rewrittenNode,
            unit: rewrittenNodeUnit,
          };
          rewrittenRules.set(node, rewrittenRule);
          return rewrittenRule;
        },
        reference(node, { state: { parentsNames }, path }) {
          const parentName = parentsNames.find((parent) =>
            availableNames.includes(`${parent} . ${node.name}`)
          );
          if (parentName) {
            node.name = `${parentName} . ${node.name}`;
          } else if (availableNames.includes(node.name)) {
          } else {
            throw Error(`Unknown reference ${node.name} in ${parentsNames[0]}`);
          }

          const associatedRule = parsedRules.rules.find(
            (rule) => rule.name === node.name
          );
          if (!inferedUnits.has(associatedRule)) {
            inferRuleUnit(associatedRule);
          }
          inferedUnits.set(node, inferedUnits.get(associatedRule));
          return node;
        },
        constant(node) {
          const type = typeof node.value;
          if (type === "boolean" || type === "string") {
            inferedUnits.set(node, { type });
          } else if (type === "number") {
            inferedUnits.set(node, { type, unit: node.unit });
          } else {
            throw new Error(`Unexpected constant type ${type}`);
          }
        },
        barème(node, { visit }) {
          const assietteUnit = inferedUnits.get(visit(node.assiette));

          const newNode = {
            ...node,
            tranches: node.tranches.map((t) => {
              if (!t.plafond) return t;
              const plafondUnit = inferedUnits.get(visit(t.plafond));
              if (plafondUnit === assietteUnit) return t;
              else {
                return {
                  ...t,
                  plafond: {
                    type: "unitConversion",
                    factor: {
                      type: "constant",
                      value: conversionFactor(assietteUnit, plafondUnit),
                    },
                    value: t.plafond,
                  },
                };
              }
            }),
          };
          inferedUnits.set(newNode, assietteUnit);
          return newNode;
        },
        produit(node, { visit }) {
          const assietteUnit = inferedUnits.get(visit(node.assiette));
          const tauxUnit = inferedUnits.get(visit(node.taux));
          inferedUnits.set(node, inferUnit("*", assietteUnit, tauxUnit));
        },
        operation(node, { visit }) {
          const leftUnit = inferedUnits.get(visit(node.left));
          const rightUnit = inferedUnits.get(visit(node.right));
          let rightUnitConvert = false;
          if (node.operator === "+" || node.operator === "-") {
            inferedUnits.set(node, leftUnit);
            rightUnitConvert = true;
          } else if (node.operator === "*" || node.operator === "/") {
            const unit = inferUnit(node.operator, leftUnit, rightUnit);
            inferedUnits.set(node, unit);
          } else if (
            node.operator === "==" ||
            node.operator === ">=" ||
            node.operator === ">" ||
            node.operator === "<=" ||
            node.operator === "<"
          ) {
            if (leftUnit.type !== rightUnit?.type) {
              throw new Error("Cannot compare different types");
            }
            inferedUnits.set(node, { type: "boolean" });
            rightUnitConvert = true;
          }

          if (rightUnitConvert && leftUnit.unit !== rightUnit.unit) {
            const newNode = {
              ...node,
              right: {
                type: "unitConversion",
                factor: {
                  type: "constant",
                  value: conversionFactor(leftUnit, rightUnit),
                },
                value: node.right,
              },
            };
            inferedUnits.set(newNode, inferedUnits.get(node));
            return newNode;
          }
        },
        ["toutes ces conditions"](node) {
          inferedUnits.set(node, { type: "boolean" });
          // TODO ensure all conditions are boolean
        },
        ["une de ces conditions"](node) {
          inferedUnits.set(node, { type: "boolean" });
          // TODO ensure all conditions are boolean
        },
        ["applicable si"](node, { visit }) {
          inferedUnits.set(node, inferedUnits.get(visit(node.value)));
        },
        variations(node, { visit, next }) {
          const firstCondition = visit(node.value[0].si);
          const firstConsequence = node.value[0].alors;
          inferedUnits.set(node, inferedUnits.get(visit(firstConsequence)));
          // ensure all test are boolean, and convert consequences to the same unit
        },
        somme(node, { visit }) {
          const firstTerm = visit(node.value[0]);
          const firstTermUnit = inferedUnits.get(firstTerm);
          inferedUnits.set(node, firstTermUnit);
        },
        plafond(node, { visit }) {
          const unit = inferedUnits.get(visit(node.value));
          inferedUnits.set(node, unit);
        },
        ["par défaut"](node, { visit }) {
          const unit = inferedUnits.get(visit(node["par défaut"]));
          inferedUnits.set(node, unit);
        },
        ["unité"](node) {
          inferedUnits.set(node, { type: "number", unit: node["unité"].value });
          // todo convert
          inferedUnits.set(node.value, {
            type: "number",
            unit: node["unité"].value,
          });
        },
        ["possibilités"](node) {
          inferedUnits.set(node, { type: "string" });
          return node;
        },
      }
    );
  }

  return walk(parsedRules, {}, { rule: (node) => inferRuleUnit(node) });
}
