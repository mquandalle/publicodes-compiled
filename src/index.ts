import { parse } from "./parser";
import { compile } from "./compiler";

export class CompiledPublicodes {
  compiledRules: any;
  cache: any = {};
  traversedRulesCache: any = {};

  constructor(source: string) {
    const jsCode = compile(parse(source));
    this.compiledRules = eval(jsCode)();
  }

  evaluate(ruleName: string) {
    const r = (x: string) => {
      this.cache[x] ??= this.compiledRules[x](r);
      return this.cache[x];
    };
    return r(ruleName);
  }

  traversedRules(ruleName: string) {
    const traversedRulesStack: string[] = [];
    const r = (x: string) => {
      if (!this.traversedRulesCache[x]) {
        const traversedRulesStartIndex = traversedRulesStack.length;
        traversedRulesStack.push(x);

        this.cache[x] = this.compiledRules[x](r);
        this.traversedRulesCache[x] = traversedRulesStack.slice(
          traversedRulesStartIndex
        );
      }
      return this.cache[x];
    };

    r(ruleName);
    return Array.from(new Set(this.traversedRulesCache[ruleName]));
  }

  resetCache() {
    this.cache = {};
    this.traversedRulesCache = {};
  }
}

export default CompiledPublicodes;
