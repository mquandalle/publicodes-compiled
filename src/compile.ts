import { walk } from "zimmerframe";
import { builders as b } from "estree-toolkit";
import { generate } from "astring";
import { type ASTPublicodesNode } from "./parse";

export function compile(parsedPublicodes: ASTPublicodesNode) {
  const state = { injectedRuntime: new Set() };
  const esTree = walk(parsedPublicodes, state, {
    publicodes(node, { visit }) {
      return b.objectExpression(
        node.rules.map((ruleNode) =>
          b.property("init", b.literal(ruleNode.name), visit(ruleNode))
        )
      );
    },

    rule(node, { visit }) {
      return b.arrowFunctionExpression([], visit(node.value));
    },

    constant(node) {
      return b.literal(node.value);
    },

    reference(node) {
      return b.callExpression(b.identifier("r"), [b.literal(node.name)]);
    },

    operation(node, { visit }) {
      return b.binaryExpression(
        node.operator,
        visit(node.left),
        visit(node.right)
      );
    },

    produit(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.assiette,
        right: node.taux,
      });
    },

    unitConversion(node, { visit }) {
      return visit({
        type: "operation",
        operator: "*",
        left: node.factor,
        right: node.value,
      });
    },

    barÃ¨me(node, { state, visit }) {
      state.injectedRuntime.add(
        "const evalBareme = (a, t) => t.map((u, i) => ((Math.max(Math.min(a, u.plafond ?? Infinity) - (t[i - 1]?.plafond ?? 0), 0)) * u.taux)).reduce((a, b) => a + b, 0);"
      );
      return b.callExpression(b.identifier("evalBareme"), [
        visit(node.assiette),
        b.arrayExpression(
          node.tranches.map((t) =>
            b.objectExpression(
              ["taux", t.plafond && "plafond"]
                .filter(Boolean)
                .map((attr) =>
                  b.property("init", b.identifier(attr), visit(t[attr]))
                )
            )
          )
        ),
      ]);
    },
  });

  const jsCodeRules = generate(esTree);
  const injectedRuntime = Array.from(state.injectedRuntime).join(";\n");
  return `((r) => {${injectedRuntime}\nreturn ${jsCodeRules}})`;
}
