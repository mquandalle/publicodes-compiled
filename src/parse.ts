import { tokenize, type Token } from "./tokenize";
import { walk } from "zimmerframe";

export type ASTNode =
  | { type: "constant"; value: number; unit?: string }
  | { type: "operation"; operator: string; left: ASTNode; right: ASTNode }
  | { type: "reference"; name: string }
  | { type: "produit"; assiette: ASTNode; taux: ASTNode }
  | {
      type: "bar√®me";
      assiette: ASTNode;
      tranches: Array<{ taux: ASTNode; plafond?: ASTNode }>;
    };

export function parse(source): Record<string, ASTNode> {
  const tokens: Token[] = tokenize(source);
  const parsedRules = {};
  let index = 0;

  while (index < tokens.length) {
    const currentToken = tokens[index];
    if (currentToken.type === "outdent") {
      index++;
      continue;
    } else if (currentToken.type === "key") {
      index++;
      const ruleName = currentToken.value;

      if (tokens[index]?.type === "key") {
        parsedRules[ruleName] = {
          type: "constant",
          value: "undefined",
        };
        continue;
      }

      parsedRules[ruleName] = parseExpression();
    } else {
      console.log(currentToken, index);
      throw new Error(`Unexpected token ${currentToken.type}`);
    }
  }

  function parseExpression() {
    switch (tokens[index]?.type) {
      case "indent":
        index++; // skip the 'indent' token
        return parseMechanism();
      case "key":
        index++;
        if (tokens[index]?.type === "indent") {
          index++;
          return parseRecord();
        } else {
          return parseInlineExpression();
        }

      case "list-item":
        return parseList();
      default:
        return parseInlineExpression();
    }
  }

  function parseList() {
    const list = [];
    while (index < tokens.length && tokens[index].type === "list-item") {
      index++;
      if (tokens[index].type === "key") {
        list.push(parseRecord());
      } else {
        list.push(parseExpression());
      }
    }
    if (index < tokens.length && tokens[index].type === "outdent") {
      index++;
    }
    return list;
  }

  function parseRecord() {
    const record = {};

    while (index < tokens.length && tokens[index].type === "key") {
      const key = tokens[index++].value;
      record[key] = parseExpression();
    }
    return record;
  }

  function parseInlineExpression(): ASTNode {
    let node = parseTerm();

    while (index < tokens.length) {
      const token = tokens[index];
      if (
        token.type === "operator" &&
        (token.value === "+" || token.value === "-")
      ) {
        index++;
        const right = parseTerm();
        node = {
          type: "operation",
          operator: token.value,
          left: node,
          right,
        };
      } else {
        break;
      }
    }
    return node;
  }

  function parseTerm(): ASTNode {
    let node = parseFactor();

    while (index < tokens.length) {
      const token = tokens[index];
      if (
        token.type === "operator" &&
        (token.value === "*" || token.value === "/")
      ) {
        index++;
        const right = parseFactor();
        node = {
          type: "operation",
          operator: token.value,
          left: node,
          right,
        };
      } else {
        break;
      }
    }
    return node;
  }

  function parseFactor(): ASTNode {
    const token = tokens[index++];
    if (token.type === "constant") {
      return { type: "constant", value: token.value, unit: token.unit };
    } else if (token.type === "reference") {
      return { type: "reference", name: token.value };
    }
    throw new Error(`Unexpected token ${JSON.stringify(token)}`);
  }

  function parseMechanism(): ASTNode {
    const mechanismName = tokens[index++];
    if (mechanismName.type !== "key") {
      throw new Error(`Unexpected token ${mechanismName}`);
    }

    const mechanismNode: ASTNode = { type: mechanismName.value } as any;

    index++; // skip the 'indent' token for mechanisms parameters

    while (index < tokens.length && tokens[index].type !== "outdent") {
      const token = tokens[index++];
      if (token.type === "key") {
        mechanismNode[token.value] = parseExpression();
      } else {
        throw new Error(`Unexpected token ${JSON.stringify(token)}`);
      }
    }

    index++; // Skip the 'outdent' token mechanisms parameters
    index++; // Skip the 'outdent' token for the mechanism itself

    return mechanismNode;
  }

  return resolveNames(parsedRules);
}

function resolveNames(parsedRules) {
  const availableNames = Object.keys(parsedRules);
  const allParents = (name) =>
    name
      .split(".")
      .reduce((acc, name) => {
        const last = acc[acc.length - 1];
        return [...acc, last ? `${last} . ${name.trim()}` : name.trim()];
      }, [])
      .reverse();
  return Object.fromEntries(
    Object.entries(parsedRules).map(([name, node]) => {
      return [
        name,
        walk(node, null, {
          reference(node) {
            const parentName = allParents(name).find((parent) =>
              availableNames.includes(`${parent} . ${node.name}`)
            );
            if (parentName) {
              return {
                type: "reference",
                name: `${parentName} . ${node.name}`,
              };
            }
            if (availableNames.includes(node.name)) {
              return node;
            }

            throw Error(`Unknown reference ${node.name} in ${name}`);
          },
        }),
      ];
    })
  );
}
