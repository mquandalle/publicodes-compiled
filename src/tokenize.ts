import { keysWithExpression } from "./parse";

export type Token =
  | { type: "number"; value: number }
  | { type: "boolean"; value: boolean }
  | { type: "text"; value: string }
  | { type: "operator"; value: string }
  | { type: "reference"; value: string }
  | { type: "unit"; value: string }
  | { type: "key"; value: string }
  | { type: "paren-open" }
  | { type: "paren-close" }
  | { type: "indent" }
  | { type: "outdent" }
  | { type: "list-item" };

const ruleNameRegex = /[\p{L}$](([$\p{L}\x20\.'0-9]|\S-\S)*[\p{L}$0-9])?/uy;
const booleanRegex = /(oui|non)/y;
const numberRegex = /[0-9]+(\.[0-9]+)?/y;
const unitRegex = /[a-z€]+(\/[a-z€]+)?/y;
const infixOperatorRegex = /(<=|>=|<|>|=|\+|\-|\*|\/)/y;
const endOfLine = /.*/y;
const spaceRegex = /\x20+/y;
const startMultiLineStringRegex = /\x20*\|\x20*\n/y;

type TokenWithLoc = Token & { start?: number; end?: number };

export function tokenize(
  source: string,
  { withLoc = false } = {}
): TokenWithLoc[] {
  let tokens: Token[] = [];
  let currentIndent = 0;
  let indentStack: number[] = [];
  let isListStack: boolean[] = [];
  let cursor = 0;
  let prevCursorPos = 0;

  function matchRegex(regex: RegExp): string | undefined {
    regex.lastIndex = cursor;
    const matched = regex.exec(source);
    if (matched) {
      if (withLoc) {
        prevCursorPos = cursor;
      }
      cursor += matched[0].length;
      return matched[0];
    }
  }
  function matchSingleChar<const C extends string>(char: C): C | undefined {
    const matched = source[cursor] === char;
    if (matched) {
      if (withLoc) {
        prevCursorPos = cursor;
      }
      cursor++;
      return char;
    }
  }
  function push(token: Token) {
    if (withLoc) {
      tokens.push({ ...token, start: prevCursorPos, end: cursor });
    } else {
      tokens.push(token);
    }
  }

  while (cursor < source.length) {
    if (cursor === 0 || matchSingleChar("\n")) {
      const spaces = matchRegex(spaceRegex);
      let indent = spaces ? spaces.length : 0;
      if (matchSingleChar("-")) {
        indent++;
        const spaces = matchRegex(spaceRegex);
        indent += spaces ? spaces.length : 0;

        if (indent > currentIndent) {
          indentStack.push(currentIndent);
          isListStack.push(true);
          currentIndent = indent;
        }
        push({ type: "list-item" });
      } else if (indent > currentIndent) {
        push({ type: "indent" });
        indentStack.push(currentIndent);
        isListStack.push(false);
        currentIndent = indent;
      } else {
        while (indent < currentIndent) {
          if (!isListStack.pop()) {
            push({ type: "outdent" });
          }
          currentIndent = indentStack.pop() || 0;
        }
      }
      if (cursor > 0) {
        continue;
      }
    }

    if (matchSingleChar("#")) {
      matchRegex(endOfLine);
      continue;
    }

    if (matchRegex(spaceRegex)) {
      continue;
    }

    const operatorToken = matchRegex(infixOperatorRegex);
    if (operatorToken) {
      push({ type: "operator", value: operatorToken });
      continue;
    }

    if (matchSingleChar("(")) {
      push({ type: "paren-open" });
      continue;
    }

    if (matchSingleChar(")")) {
      push({ type: "paren-close" });
      continue;
    }

    const booleanToken = matchRegex(booleanRegex);
    if (booleanToken) {
      push({ type: "boolean", value: booleanToken === "oui" });
      continue;
    }

    let quoteType = matchSingleChar('"') || matchSingleChar("'");
    if (quoteType) {
      let value = "";
      let legacyDuplicatedQuotes = false;

      const invertQuote = (q: '"' | "'") => (q === '"' ? "'" : '"');
      if (matchSingleChar(invertQuote(quoteType))) {
        legacyDuplicatedQuotes = true;
        quoteType = invertQuote(quoteType);
      }

      for (; cursor < source.length; cursor++) {
        if (source[cursor] === "\\") {
          value += source[++cursor];
          continue;
        } else if (source[cursor] === quoteType) {
          push({ type: "text", value });
          if (withLoc) {
            tokens[tokens.length - 1].end += 1;
          }
          cursor++;

          if (legacyDuplicatedQuotes) {
            if (!matchSingleChar(invertQuote(quoteType))) {
              throw new Error("Unterminated legacy string");
            }
          }
          break;
        }

        value += source[cursor];
      }

      if (cursor > source.length) {
        throw new Error("Unterminated string");
      }

      continue;
    }
    const refToken = matchRegex(ruleNameRegex);
    if (refToken) {
      matchRegex(spaceRegex);
      if (matchSingleChar(":")) {
        push({ type: "key", value: refToken });
        isListStack[isListStack.length - 1] = false;
        if (matchRegex(startMultiLineStringRegex)) {
          multiLineString();
        } else if (refToken === "unité") {
          matchRegex(spaceRegex);
          const unit = matchRegex(unitRegex);
          if (unit) {
            push({ type: "unit", value: unit });
          } else {
            throw new Error("Missing unit");
          }
        } else if (
          !keysWithExpression.has(refToken) &&
          indentStack.length > 0
        ) {
          matchRegex(spaceRegex);
          const text = matchRegex(endOfLine);
          if (text) {
            push({ type: "text", value: text });
          }
        }
      } else {
        push({ type: "reference", value: refToken });
      }
      continue;
    }

    const numberToken = matchRegex(numberRegex);
    if (numberToken) {
      matchRegex(spaceRegex);
      const percentage = matchSingleChar("%");
      const value = parseFloat(numberToken);
      if (percentage) {
        push({ type: "number", value: value / 100 });
      } else {
        push({ type: "number", value });
      }
      const unit = matchRegex(unitRegex);
      if (unit) {
        push({ type: "unit", value: unit });
      }
      continue;
    }
    throw new Error(`Cannot tokenize`);
  }

  function multiLineString() {
    const endMultiLineString = RegExp(
      "(.+?)\\n\\x20{0," + currentIndent + "}([^\\x20\\n]|$)",
      "ys"
    );
    endMultiLineString.lastIndex = cursor;
    const multilineString = endMultiLineString.exec(source);

    if (!multilineString) {
      throw new Error("Unterminated multi-line string");
    }
    const length = multilineString[1].length;
    const value = source
      .slice(cursor, cursor + length)
      .replace(/(^|\n)\x20*/g, "\n")
      .trim();
    cursor += multilineString[1].length;
    push({ type: "text", value });
  }

  return tokens;
}

export function printTokens(tokens: Token[], { lineNumbers = false } = {}) {
  return tokens
    .map(
      (t, i) =>
        `${
          lineNumbers
            ? i.toString().padEnd(tokens.length.toString().length + 1, " ")
            : ""
        }${t.type}${"value" in t ? `(${t.value})` : ""}`
    )
    .join("\n");
}
