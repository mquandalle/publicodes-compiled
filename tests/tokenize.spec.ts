import { expect, test } from "bun:test";
import { tokenize, printTokens } from "../src/tokenize";

const snap = (intput: string) => printTokens(tokenize(intput));

test("Tokenize", () => {
  const canonicalStringDeclaration = "string: 'hello'";
  expect(snap(`string: 'hello'`)).toMatchSnapshot();
  expect(snap(`string: "hello"`)).toEqual(snap(canonicalStringDeclaration));
  expect(snap(`string: "'hello'"`)).toEqual(snap(canonicalStringDeclaration));
  expect(snap(`string: '"hello"'`)).toEqual(snap(canonicalStringDeclaration));

  expect(snap(`chiffre affaires: 20000 €/mois`)).toMatchSnapshot();
  expect(snap(`localisation . ZFE: oui`)).toMatchSnapshot();
  expect(snap(`âge >= 6`)).toMatchSnapshot();
});
