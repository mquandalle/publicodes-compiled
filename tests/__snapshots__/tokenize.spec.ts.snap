// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenize 1`] = `
"key(string)
text(hello)"
`;

exports[`Tokenize 2`] = `
"key(chiffre affaires)
constant(20000) [unit: â‚¬/mois]"
`;

exports[`Tokenize 3`] = `
"key(localisation . ZFE)
boolean(true)"
`;

exports[`Tokenize 4`] = `
"reference(Ã¢ge)
operator(>=)
constant(6)"
`;
